libfprint/drivers/crfpmoc/crfpmoc.c

```diff
@@ -0,0 +1,627 @@
+                  fpi_device_identify_report (device, NULL, NULL, fpi_device_retry_new_msg (FP_DEVICE_RETRY_GENERAL, "FP mode: (0x%x)", mode));
+                  fpi_device_identify_complete (device, NULL);
+                }
+              fpi_ssm_mark_completed (ssm);
+            }
+        }
+      break;
+    }
+}
+
+static void
+crfpmoc_identify_verify (FpDevice *device)
+{
+  fp_dbg ("Identify or Verify");
+  FpiDeviceCrfpMoc *self = FPI_DEVICE_CRFPMOC (device);
+
+  g_assert (self->task_ssm == NULL);
+  self->task_ssm = fpi_ssm_new (device, crfpmoc_verify_run_state, VERIFY_STATES);
+  fpi_ssm_start (self->task_ssm, crfpmoc_task_ssm_done);
+}
+
+static void
+crfpmoc_clear_storage_run_state (FpiSsm *ssm, FpDevice *device)
+{
+  FpiDeviceCrfpMoc *self = FPI_DEVICE_CRFPMOC (device);
+  int r;
+  guint32 mode;
+  const gchar *error_msg;
+
+  switch (fpi_ssm_get_cur_state (ssm))
+    {
+    case CLEAR_STORAGE_SENSOR_RESET:
+      r = crfpmoc_cmd_fp_mode (self, CRFPMOC_FP_MODE_RESET_SENSOR, &mode, &error_msg);
+      if (r < 0)
+        {
+          fpi_ssm_mark_failed (ssm, fpi_device_error_new_msg (FP_DEVICE_ERROR_GENERAL, "%s", error_msg));
+        }
+      else
+        {
+          fpi_device_clear_storage_complete (device, NULL);
+          fpi_ssm_mark_completed (ssm);
+        }
+      break;
+    }
+}
+
+static void
+crfpmoc_clear_storage (FpDevice *device)
+{
+  fp_dbg ("Clear storage");
+  FpiDeviceCrfpMoc *self = FPI_DEVICE_CRFPMOC (device);
+
+  g_assert (self->task_ssm == NULL);
+  self->task_ssm = fpi_ssm_new (device, crfpmoc_clear_storage_run_state, CLEAR_STORAGE_STATES);
+  fpi_ssm_start (self->task_ssm, crfpmoc_task_ssm_done);
+}
+
+static void
+crfpmoc_list_run_state (FpiSsm *ssm, FpDevice *device)
+{
+  g_autoptr(GPtrArray) enrolled_prints = NULL;
+
+  switch (fpi_ssm_get_cur_state (ssm))
+    {
+    case LIST_RETURN_ENROLLED_PRINTS:
+      enrolled_prints = gallery_data_load (device);
+      fpi_device_list_complete (device, g_steal_pointer (&enrolled_prints), NULL);
+      fpi_ssm_mark_completed (ssm);
+      break;
+    }
+}
+
+static void
+crfpmoc_list (FpDevice *device)
+{
+  fp_dbg ("List");
+  FpiDeviceCrfpMoc *self = FPI_DEVICE_CRFPMOC (device);
+
+  g_assert (self->task_ssm == NULL);
+  self->task_ssm = fpi_ssm_new (device, crfpmoc_list_run_state, LIST_STATES);
+  fpi_ssm_start (self->task_ssm, crfpmoc_task_ssm_done);
+}
+
+static void
+fpi_device_crfpmoc_init (FpiDeviceCrfpMoc *self)
+{
+  G_DEBUG_HERE ();
+  self->fd = -1;
+}
+
+static void
+fpi_device_crfpmoc_class_init (FpiDeviceCrfpMocClass *klass)
+{
+  FpDeviceClass *dev_class = FP_DEVICE_CLASS (klass);
+
+  dev_class->id = FP_COMPONENT;
+  dev_class->full_name = CRFPMOC_DRIVER_FULLNAME;
+
+  dev_class->type = FP_DEVICE_TYPE_UDEV;
+  dev_class->scan_type = FP_SCAN_TYPE_PRESS;
+  dev_class->id_table = crfpmoc_id_table;
+  dev_class->nr_enroll_stages = CRFPMOC_NR_ENROLL_STAGES;
+  dev_class->temp_hot_seconds = 0;
+
+  dev_class->open = crfpmoc_open;
```

Comment:
this device has a peculiar set of features which doesn't correspond to traditional MoC drivers.

enrollment works but we can't associate data with it, we can know which finger (as in the the print number - 0, 1, etc.) though by checking the number of enrolled fingers

instead of verify / identify, there is a "match" feature - if you press your finger, it can tell if it matches one of the enrolled ones. We also get info on which finger it was that matched, so I guess it's like "identify" but without the associated data.

there is no delete individual print feature (actually, probably there is but not apparent) - there is one to delete all registered prints though. 

due to these peculiarities, I've implemented a "hybrid" approach where the print data is stored in a file instead of on device - does this sound like a viable approach?